Tile Reskinning: Allows you to make prettier in-game building projects

KathrinBailey opened the pull request:
## About The Pull Request
• Allows players to shift-click on a stack of titanium tiles to change which type of titanium tile they wish to place. We have five. Default, orange, purple, white and blue.
• Reintroduces the old titanium tiles back, available through tile reskinning. They can be placed in the editor without icon_state varediting.
• Introduces """new""" tile stack sprites for titanium (they're just color changes to the old sprites).

## Why It's Good For The Game
This allows you to make builds that otherwise are impossible outside of requesting an admin to varedit for you. Giving the players more freedom to do prettier builds makes engineering a considerably more fun job. Part of why I play engineer is to do silly builds, like build a new medbay, or build a "party shuttle" (Beestation custom whiteships didn't exist yet). One thing I wish I had then was more freedom to make things look prettier. Whilst this PR is, at the moment, only doing it for titanium tiles, it's definitely a start.
![image](https://user-images.githubusercontent.com/53862927/95030164-a60a5e80-06a5-11eb-9c7a-61108d464226.png)
Here's the tiles you can place in-game.
**Remember when Beestation made custom whiteships possible? Here's a custom whiteship.**
![image](https://user-images.githubusercontent.com/53862927/95030394-5b89e180-06a7-11eb-83e5-cd0c338169f0.png)
Whoo, you can do that in-game.

## For Maintainers
[I'm porting my own PR from elsewhere here](https://github.com/Skyrat-SS13/Skyrat13/pull/2838), I'd like to port the full thing but I'm unsure on how you'd want it, so I've left it out for now.
Although the method used for titanium doesn't work with any floors that are a subtype of plasteel, I can easily allow players to do this through a slightly different method. Here's what it looks like.
![image](https://user-images.githubusercontent.com/53862927/95030181-d81bc080-06a5-11eb-8129-9be2abcdaa66.png) (Ignore the stack tile icons, I made them before I realised I needed to do this another way)
Unfortunately because of how icon_regular_floor (I think that was the pesky var?) works, it doesn't matter if I try and do things like I've done for titanium. It won't work. So instead, the original PR just allowed players to whack a plasteel floor subtype with plasteel tiles and they could just change the icon_state.
Please let me know if:
• You don't want players to change plasteel and subtypes of plasteel into the different subtypes.
• You want players to be able to do this, but through a different method than whacking (clicking )the floor with a plasteel tile stack.
• You want players to be able to do this and you are fine with players doing this by poking the floor with a plasteel tile stack.

## Changelog
:cl: Floof Ball#0798
add: The ability to change what type of titanium floor tiles you wish to place via shift-clicking your stack of titanium floor tiles.
/:cl:
_Whilst the screenshots taken here are not from TG, I've tested this and it indeed works just fine on TG._

KathrinBailey committed:
Tile Floofing port

https://github.com/Skyrat-SS13/Skyrat13/pull/2838

Jared-Fogle commented on code/game/objects/items/stacks/tiles/tile_reskinning.dm:
```diff
@@ -0,0 +1,54 @@
+// A: Unfortunately, this does not work on subtypes of plasteel and instead we must change the icon_state of these turfs instead, as the icon_regular_floor var that "saves" what type of floor a plasteel subtype turf was so once repaired...
+// ... it'll go back to the floor it was instead of grey (medical floors turn white even after crowbaring the tile and putting it back). This stops changing turf_type from working.
+
+/obj/item/stack/tile/mineral/titanium/ShiftClick(mob/user)
```
Does anything else use shift click like this? I understand that alt-click splits the stack, but I'm not sure if overriding shift click is common.

KathrinBailey replied:
Shift click should just examine, I'm pretty sure it doesn't interfere with anything else.

Rohesie replied:
Does self-attack do anything here? Clicking it while in your hand, or pressing Z.
Might be a bit more intuitive than shift click, which is usually for examine instead.

KathrinBailey replied:
Self-attack will bring up the crafting menu for items you craft with titanium floor tiles. There's nothing to actually craft of course.

Jared-Fogle replied:
Sounds like you should replace that then :eyes:

Fikou replied:
why not ctrl click when holding

Jared-Fogle commented on code/game/objects/items/stacks/tiles/tile_reskinning.dm:
```diff
@@ -0,0 +1,54 @@
+// ... it'll go back to the floor it was instead of grey (medical floors turn white even after crowbaring the tile and putting it back). This stops changing turf_type from working.
+
+/obj/item/stack/tile/mineral/titanium/ShiftClick(mob/user)
+	var/choice = input(user, "What tile would you like?") as null|anything in list("Titanium", "Yellow Titanium", "Blue Titanium", "White Titanium", "Purple Titanium", "Titanium Tile", "Yellow Titanium Tile", "Blue Titanium Tile", "White Titanium Tile", "Purple Titanium Tile")
```
Are you kind enough to make this a radial so I don't have to? 🙂

KathrinBailey replied:
Sure! I haven't touched anything regarding radial menus; the thought of radial menus didn't ever cross my mind but I'll look into other implementations of radial menus and I'll try and learn how to make it work for this.
I suppose DNM for now, it's learning time.

ArcaneMusic replied:
The good news for a usecase like this is that it's a static radial, no need to come up with a dynamic radial option. Look at things like the I did with TCG cards for an example:
https://github.com/tgstation/tgstation/blob/2844c95561c81bff1b7cf2fc10d0503d32824112/code/game/objects/items/tcg/tcg.dm#L52

KathrinBailey replied:
![image](https://user-images.githubusercontent.com/53862927/95046030-2eefbd00-06db-11eb-9ca5-bfe899fe099d.png)
@Jared-Fogle 
Okay so it's working, but I think it can be improved further. Is there a way to set the amount of options per page of the radial menu?
There are two tiles, each tiles has 5 variants, if I could set it to only have 5 options per page, I can have one tile and it's variants only on each page, instead of cramming 7 options on one page and 3 on the other.

Jared-Fogle replied:
> Is there a way to set the amount of options per page of the radial menu?

Increase the radius parameter when calling show_radial_menu.

KathrinBailey replied:
Radius does urh, I guess I should've expected radius increases the radius between the centre and buttons.
![image](https://user-images.githubusercontent.com/53862927/95047787-99562c80-06de-11eb-9945-94a9ad6449af.png)
It doesn't change how many buttons per page there are, heck.

Dennok replied:
What if make 3 menu, first has 2 variants of tiles, and another 2 has 5 variants of color.

Jared-Fogle replied:
@KathrinBailey Haven't done it myself, just code diving. I suppose if you can't figure it out, just leave it as is and in the future we can add a new parameter to show_radial_menu that lets you do this.

ArcaneMusic commented:
Honestly, very hot PR, I'm a fan.

KathrinBailey committed:
Radial menus

It's not yet as good as it could be but it works.

Jared-Fogle commented on code/game/objects/items/stacks/tiles/tile_reskinning.dm:
```diff
@@ -0,0 +1,70 @@
+	var/list/choices = GLOB.titanium_tile_reskinning_radial_choices
+	if(!length(choices))
+		choices = GLOB.titanium_tile_reskinning_radial_choices = list(
+		"Titanium" 				= image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle"),
+		"Yellow Titanium" 		= image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_yellow"),
+		"Blue Titanium" 		= image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_blue"),
+		"White Titanium" 		= image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_white"),
+		"Purple Titanium" 		= image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_purple"),
+		"Titanium Tile" 		= image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_old"),
+		"Yellow Titanium Tile" 	= image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_old_yellow"),
+		"Blue Titanium Tile" 	= image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_old_blue"),
+		"White Titanium Tile" 	= image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_old_white"),
+		"Purple Titanium Tile" 	= image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_old_purple"),
```
```suggestion
			"Titanium" = image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle"),
			"Yellow Titanium" = image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_yellow"),
			"Blue Titanium" = image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_blue"),
			"White Titanium" = image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_white"),
			"Purple Titanium" = image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_purple"),
			"Titanium Tile" = image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_old"),
			"Yellow Titanium Tile" = image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_old_yellow"),
			"Blue Titanium Tile" = image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_old_blue"),
			"White Titanium Tile" = image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_old_white"),
			"Purple Titanium Tile" = image(icon = 'icons/obj/tiles.dmi', icon_state = "tile_shuttle_old_purple"),
```

KathrinBailey committed:
Update code/game/objects/items/stacks/tiles/tile_reskinning.dm

Co-authored-by: Jared-Fogle <35135081+Jared-Fogle@users.noreply.github.com>

Rohesie commented on code/game/objects/items/stacks/tiles/tile_reskinning.dm:
```diff
@@ -0,0 +1,70 @@
+// A: Unfortunately, this does not work on subtypes of plasteel and instead we must change the icon_state of these turfs instead, as the icon_regular_floor var that "saves" what type of floor a plasteel subtype turf was so once repaired...
+// ... it'll go back to the floor it was instead of grey (medical floors turn white even after crowbaring the tile and putting it back). This stops changing turf_type from working.
+
+GLOBAL_LIST_EMPTY(titanium_tile_reskinning_radial_choices)
+
+/obj/item/stack/tile/mineral/titanium/attack_self(mob/user)
+	var/list/choices = GLOB.titanium_tile_reskinning_radial_choices
+	if(!length(choices))
+		choices = GLOB.titanium_tile_reskinning_radial_choices = list(
```
```suggestion
/obj/item/stack/tile/mineral/titanium/attack_self(mob/user)
	var/static/list/choices = list(
```
Since this won't really be used by anything else you can just make it a proc-exclusive static list.
For reference, this will be only be populated the first time the proc is called, and re-used on further proc calls.

Rohesie commented on code/game/turfs/open/floor/mineral_floor.dm:
```diff
@@ -119,6 +119,39 @@
+// OLD TITANIUM (titanium floor tiles before PR #50454)
+/turf/open/floor/mineral/titanium/old
+	name = "titanium tile"
+	icon_state = "titanium_old"
```
why the `old` part in the name?

KathrinBailey replied:
It's called "old" because it's the old sprites returning back into the game.
It's just called that because I can't think of a better name that lets mappers/coders know what it is better than "it's the old titanium floor sprite".

KathrinBailey replied:
They're called "tiled" now.
![image](https://user-images.githubusercontent.com/53862927/95148278-d7a12980-077a-11eb-803e-759490c0aaba.png)
It's kinda... tile-y?

Rohesie replied:
Can you also change the `icon_state` names to match?

KathrinBailey committed:
Rohesie's suggested change.

This title autocompletes after putting it so many times. x_x

Co-authored-by: Rohesie <rohesie@gmail.com>

KathrinBailey committed:
Name change for Rohesie

Typepath was called old because it's the old sprite, now it's tiled because Rohesie correctly pointed out that "old" changes fast.

KathrinBailey committed:
Increases size of radial menu

Looks neater

KathrinBailey commented:
I think that's everything, I gave it another test and it works.

Qustinnus commented:
optional but might be cool to make some of these craftable from other materials? All of them being titanium kinda kills reasons to use other floors, for example some of those tiles look more like gold floors than gold floors do.

Jared-Fogle commented:
Is this still WIP?

KathrinBailey commented:
> Is this still WIP?

Nah sorry, forgot to change the title.
I wanna add more tiles to reskin for different types, but I can do that in a separate PR.

Jared-Fogle commented on code/game/objects/items/stacks/tiles/tile_mineral.dm:
```diff
@@ -71,7 +71,7 @@
 	name = "titanium tile"
 	singular_name = "titanium floor tile"
-	desc = "A tile made of titanium, used for shuttles."
+	desc = "A tile made of titanium, used for shuttles. Shift-click to change what type of titanium tiles you want."
```
```suggestion
	desc = "A tile made of titanium, used for shuttles. Use while in hand to change what type of titanium tiles you want."
```

KathrinBailey committed:
Update code/game/objects/items/stacks/tiles/tile_mineral.dm

Co-authored-by: Jared-Fogle <35135081+Jared-Fogle@users.noreply.github.com>
